radix-engine/src/engine/call_frame.rs

```diff
@@ -1708,27 +1741,53 @@ where
         );
 
         // invoke the main function
-        let (result, received_values) = frame.run(snode_ref, loaded_snode, &fn_ident, input)?;
+        let (mut result, received_values) = frame.run(snode_ref, loaded_snode, &fn_ident, input)?;
         drop(frame);
 
         // Release locked addresses
         for l in locked_values {
-            self.track.release_lock(l);
+            // TODO: refactor after introducing `Lock` representation.
+            self.track
+                .release_lock(l.clone(), is_pay_fee && matches!(l, Address::Vault(..)));
         }
 
-        // move buckets and proofs to this process.
-        for (id, value) in received_values {
-            trace!(self, Level::Debug, "Received value: {:?}", value);
-            self.owned_values.insert(id, value);
+        if is_pay_fee {
```

Comment:
Not sure why this part is here instead of in the method itself?

Reply:
Originally trying to avoid overloading the `SystemApi` interface to support fee accounting, but might be a good idea to have the logic within `Vault` model.

Will refactor.